### PR TITLE
Sequence task (1) Semenyuk

### PR DESCRIPTION
**Pull request by VladSemenyuk**

### CODE REVIEW

File: sequence_task_1/student/include/impl.h [hunk 1]
```diff
@@ -1,51 +1,86 @@
...
+  using size_type = typename Container::size_type;
+  using value_type = typename Container::value_type;
+  using reference = typename Container::reference;
+  using const_reference = typename Container::const_reference;
+  using iterator = typename Container::iterator;
+  using const_iterator = typename Container::const_iterator;
+  using difference_type = typename Container::difference_type;
+
+  /** @todo Iterators */
+  const_iterator begin() {
```
Comment (smay1613) [inline sequence_task_1/student/include/impl.h:27]: Mark as const

File: sequence_task_1/student/include/impl.h [hunk 2]
```diff
@@ -1,51 +1,86 @@
...
-    /** @todo Iterators */
+  const_iterator end() {
+    return m_tracklist.end();
+  }
 
-    /** @todo Constructor from any reversible sequence container */
+  /** @todo Constructor from any reversible sequence container */
+  template <class T>
+  StaticPlaylist(const T& track_list) {
+    m_tracklist.assign(track_list.begin(), track_list.end());
```
Comment (smay1613) [inline sequence_task_1/student/include/impl.h:38]: Use initialization list and range constructor

File: sequence_task_1/student/include/impl.h [hunk 3]
```diff
@@ -1,51 +1,86 @@
...
+    return *this;
+  }
 
-    /** @todo Add track from initializer */
-    template<class... Args>
-    const Song& play(Args&&... songData);
+  /** @todo Add track from initializer */
+  template <class... Args>
+  const Song& play(Args&&... songData) {
+    m_tracklist.emplace_back(songData...);
```
Comment (smay1613) [inline sequence_task_1/student/include/impl.h:51]: Use std::forward
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#f19-for-forward-parameters-pass-by-tp-and-only-stdforward-the-parameter

File: sequence_task_1/student/include/impl.h [hunk 4]
```diff
@@ -1,51 +1,86 @@
...
+  }
 
-    /** @todo Skip to the next track in playlist, remove current */
-    void switchNext();
+  /** @todo Skip to the next track in playlist, remove current */
+  void switchNext() {
+    if (m_tracklist.empty()) {
+      throw std::out_of_range("");
+    }
+    m_tracklist.erase(m_tracklist.begin());
```
Comment (smay1613) [inline sequence_task_1/student/include/impl.h:71]: **Not effective for vector!**

File: sequence_task_1/student/include/impl.h [hunk 5]
```diff
@@ -1,51 +1,86 @@
...
+  /** @todo Get first track in playlist stack */
+  const Song& current() const {
+    return m_tracklist.front();
+  }
 
-    /** @todo Skip to the next track in playlist, remove current */
-    void switchNext();
+  /** @todo Skip to the next track in playlist, remove current */
+  void switchNext() {
+    if (m_tracklist.empty()) {
```
Comment (smay1613) [inline sequence_task_1/student/include/impl.h:68]: Reuse existing methods

**Review by smay1613 — CHANGES_REQUESTED** [review]

File: sequence_task_1/student/include/impl.h (hunk 4 above)
Comment (VladSemenyuk) [reply sequence_task_1/student/include/impl.h:70]: Can't find a way to do this more efficiently without losing the songs order.

Comment (smay1613) [reply sequence_task_1/student/include/impl.h:70]: What about the reverse iterators? :)
Maybe we need a special way to copy the input data.

**Pull request closed without merging**